File: 454-4sum-ii/4sum-ii.ts
function fourSumCount(nums1: number[], nums2: number[], nums3: number[], nums4: number[]): number {
    const sumMap: Map<number, number> = new Map()
    let count: number = 0

    for (let a of nums1) {
        for (let b of nums2) {
            const sum: number = a + b
            sumMap.set(sum, (sumMap.get(sum) || 0) + 1)
        }
    }

    for (let c of nums3) {
        for (let d of nums4) {
            const target = -(c + d)
            if (sumMap.has(target)) {
                count += sumMap.get(target)
            }
        }
    }

    return count
};

File: 454-4sum-ii/4sum-ii.py
class Solution:
    def fourSumCount(self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]) -> int:
        sum_map = defaultdict(int)

        for a in nums1:
            for b in nums2:
                current_sum = a + b
                sum_map[current_sum] += 1

        count = 0

        for c in nums3:
            for d in nums4:
                target = -(c + d)
                if target in sum_map:
                    count += sum_map[target]

        return count